README

Games::Go::AGA::BayRate:
------------------------

The American Go Association (AGA) provides a rating system for the go
players of the nation.  The algorithm is described in detail in a paper
on their web-site (http://usgo.org):  AGARatings-Math.pdf.  They also
provide a C++ implementation example: bayrate.zip.

This module implements a perl version of bayrate, both the executable
and the support objects (Game, Player, etc).  Note: only bayrate is
included here, singlerate and check are left as an excersise for the
student.

bayrate, and this module, require a fairly recent version of GNU Scientific
Library (GSL), including the devel portion (containing header files, etc).
If you are using Fedora (linux), "yum install gsl gsl-devel" should be
sufficient.  Version 1.14 and later should work, earlier versions may
not.  To find your current version, run:

    pkg-config --modversion gsl

Work is being done to provide a full perl interface to GSL, but it is
not complete as of this writing.  I have used Inline::C to hook to the
specific GSL functions used by bayrate.pl.  The missing parts needed by
bayrate.pl are the Multimin functions (f_minimizer and fdf_minimizer).
Tests for the Inline::C interface to these functions (as well as the C
versions as called out in the Gnu GSL documentation) are included in the
'extra' subdirectory of this package.


INSTALL:
--------

    perl Makefile.PL
    make
    make test
    make install

COPYRIGHT:
---------

    
This software is copyright (c) 2011 by Reid Augustin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Perl extension Games::Go::AGA::BayRate

0.071     2014-01-21 08:59:32 America/Los_Angeles
    Allow set_crating to have a value of 0

0.066     2013-12-04 15:00:52 America/Los_Angeles
    Check for GSL libs using pkg-config or gsl-config instead
        of Devel::CheckLib.
    Modified G::G::A::B::GSL::Multimin.pm to use pkg-config or
        gsl-config.  this make GSL install via Alien::GSL work.

0.064     2013-10-05
    Original release.



extra/f_test.pl
#!/usr/bin/perl
#===============================================================================
#
#     ABSTRACT: Implement the example at http://www.gnu.org/software/gsl/manual/html_node/Multimin-Examples.html
#
#       AUTHOR:  Reid Augustin
#        EMAIL:  reid@LucidPort.com
#      CREATED:  05/25/2011 09:41:15 AM
#===============================================================================

use strict;
use warnings;

# use Math::GSL qw( :all );
use Math::GSL::Vector qw( :all );
use Math::GSL::Errno  qw( :all );           # for GSL_SUCCESS and GSL_CONTINUE
use Math::GSL::Multimin (qw( :all ));       # for minimizer types (conjugate_fr)
use Games::Go::AGA::BayRate::GSL::Multimin; # use this hack instead of Math::GSL::Multimin

use Readonly;

# param indices:
Readonly my $PAR_CENTER_X => 0;
Readonly my $PAR_CENTER_Y => 1;
Readonly my $PAR_SCALE_X  => 2;
Readonly my $PAR_SCALE_Y  => 3;
Readonly my $PAR_MINIMUM  => 4;

# debug aid
sub print_vector {
    my ($raw_v, $count) = @_;

    if (not $count or $count <= 0) {
        $count = Games::Go::AGA::BayRate::GSL::Multimin::my_vector_size($raw_v);    # what a hack!
    }
    my $ii;
    for ($ii = 0; $ii < $count; $ii ++) {
        if ($ii % 10 == 0) { printf("%3d:", $ii); }
        printf(" % .24g", gsl_vector_get($raw_v, $ii));
        if ($ii % 10 == 9) { print("\n"); }
    }
    if ($ii % 10 != 0) { print("\n"); }
}

# return the value of the function (specified by $params) at $v
sub my_f {
    my ($raw_v, $params) = @_;

print("my_f v:\n"); print_vector($raw_v);
    my $x = gsl_vector_get($raw_v, 0);
    my $y = gsl_vector_get($raw_v, 1);
    my $ret = $params->[$PAR_SCALE_X] * ($x - $params->[$PAR_CENTER_X]) * ($x - $params->[$PAR_CENTER_X])
            + $params->[$PAR_SCALE_Y] * ($y - $params->[$PAR_CENTER_Y]) * ($y - $params->[$PAR_CENTER_Y])
            + $params->[$PAR_MINIMUM];
printf("my_f returns % .24g,  v:\n", $ret); print_vector($raw_v);
    return $ret;
}

sub minimize {

    # Paraboloid center at (1,2), scale factors (10, 20),
    #   minimum value 30
    my @params = (
        1.0,    # $PAR_CENTER_X
        2.0,    # $PAR_CENTER_Y
        10.0,   # $PAR_SCALE_X
        20.0,   # $PAR_SCALE_Y
        30.0    # $PAR_MINIMUM
    );

    # Starting point (5, 7)
    my $x = Math::GSL::Vector->new([5.0, 7.0]);
    my $raw_x = $x->raw;
    # step size (1.0, 1.0)
    my $ss = Math::GSL::Vector->new([1.0, 1.0]);
    my $raw_ss = $ss->raw;

    # minimizer 'state'
    my $state = Games::Go::AGA::BayRate::GSL::Multimin::my_fminimizer_set (
        $gsl_multimin_fminimizer_nmsimplex,    # minimizer type
        # gsl_multimin_function_f structure members:
            \&my_f,     # f       function
            2,          # n       number of free variables
            \@params,   # params  function params passed to f
        # end of gsl_multimin_function_f structure members:
        $x->raw,    # starting point vector
        $ss->raw,   # step size
     );

    my $iter = 0;
    my $status = $GSL_CONTINUE;

    while(     $status == $GSL_CONTINUE
           and $iter <= 100) {
        $iter++;
        $status = gsl_multimin_fminimizer_iterate($state);

        last if ($status);

        my $size = gsl_multimin_fminimizer_size ($state);
        $status = gsl_multimin_test_size($size, 1e-2);

        if ($status == $GSL_SUCCESS) {
            print "\nConverged to minimum at\n";
        }
        my $raw_x = gsl_multimin_fminimizer_x($state);
        printf("%5d %10.3e %10.3e f() = %7.3f size = %.3f\n",
            $iter,
            gsl_vector_get($raw_x, 0),
            gsl_vector_get($raw_x, 1),
          # gsl_multimin_fminimizer_fval($state),    # hmm, struct member, not a function
                                                    # ok, do it this way instead:
            Games::Go::AGA::BayRate::GSL::Multimin::my_fminimizer_fval($state),
            $size,
        );
    }

    if ($status != $GSL_SUCCESS) {
        print "Error: ", gsl_strerror($status), "\n";
    }

    return $status;
}

minimize();


extra/fdf_test.pl
#!/usr/bin/perl
#===============================================================================
#
#     ABSTRACT: Implement the example at http://www.gnu.org/software/gsl/manual/html_node/Multimin-Examples.html
#
#       AUTHOR:  Reid Augustin
#        EMAIL:  reid@LucidPort.com
#      CREATED:  05/25/2011 09:41:15 AM
#===============================================================================

use strict;
use warnings;

# use Math::GSL qw( :all );
use Math::GSL::Vector qw( :all );
use Math::GSL::Errno  qw( :all );           # for GSL_SUCCESS and GSL_CONTINUE
use Math::GSL::Multimin (qw( :all ));       # for minimizer types (conjugate_fr)
use Games::Go::AGA::BayRate::GSL::Multimin; # use this hack instead of Math::GSL::Multimin

use Readonly;

# param indices:
Readonly my $PAR_CENTER_X => 0;
Readonly my $PAR_CENTER_Y => 1;
Readonly my $PAR_SCALE_X  => 2;
Readonly my $PAR_SCALE_Y  => 3;
Readonly my $PAR_MINIMUM  => 4;

# debug aid
sub print_vector {
    my ($raw_v, $count) = @_;

    if (not $count or $count <= 0) {
        $count = Games::Go::AGA::BayRate::GSL::Multimin::my_vector_size($raw_v);    # what a hack!
    }
    my $ii;
    for ($ii = 0; $ii < $count; $ii ++) {
        if ($ii % 10 == 0) { printf("%3d:", $ii); }
        printf(" % .24g", gsl_vector_get($raw_v, $ii));
        if ($ii % 10 == 9) { print("\n"); }
    }
    if ($ii % 10 != 0) { print("\n"); }
}

# return the value of the function (specified by $params) at $v
sub my_f {
    my ($raw_v, $params) = @_;

# print("my_f v:\n"); print_vector($raw_v);
    my $x = gsl_vector_get($raw_v, 0);
    my $y = gsl_vector_get($raw_v, 1);
    my $ret = $params->[$PAR_SCALE_X] * ($x - $params->[$PAR_CENTER_X]) * ($x - $params->[$PAR_CENTER_X])
            + $params->[$PAR_SCALE_Y] * ($y - $params->[$PAR_CENTER_Y]) * ($y - $params->[$PAR_CENTER_Y])
            + $params->[$PAR_MINIMUM];
# printf("my_f returns % .24g,  v:\n", $ret);
# print_vector($raw_v);
    return $ret;
}

# The value of the first derivative of f, df = (df/dx, df/dy) at $v
sub my_df {
    my ($raw_v, $params, $raw_df) = @_;

# print("my_df: v:\n"); print_vector($raw_v);
# print("my_df: df:\n"); print_vector($raw_df);
    my $x = gsl_vector_get($raw_v, 0);
    my $y = gsl_vector_get($raw_v, 1);

    gsl_vector_set($raw_df, 0, 2.0 * $params->[$PAR_SCALE_X] * ($x - $params->[$PAR_CENTER_X]));
    gsl_vector_set($raw_df, 1, 2.0 * $params->[$PAR_SCALE_Y] * ($y - $params->[$PAR_CENTER_Y]));
# print("my_df returns df:\n"); print_vector($raw_df);
}

# Compute both f and df together. set ${$f} to the value of f and $df to the
# value of the derivative
sub my_fdf {
    my ($raw_x, $params, $f, $raw_df) = @_;

# print("my_fdf: v:\n"); print_vector($raw_x);
# print("my_fdf: df:\n"); print_vector($raw_df);
    ${$f} = my_f($raw_x, $params);
    my_df($raw_x, $params, $raw_df);
# printf("my_fdf returns f=% .24g, df:\n", ${$f} ); print_vector($raw_df);
}

sub minimize {

    # Paraboloid center at (1,2), scale factors (10, 20),
    #   minimum value 30
    my @params = (
        1.0,    # $PAR_CENTER_X
        2.0,    # $PAR_CENTER_Y
        10.0,   # $PAR_SCALE_X
        20.0,   # $PAR_SCALE_Y
        30.0    # $PAR_MINIMUM
    );

    # Starting point (5, 7)
    my $x = Math::GSL::Vector->new([5.0, 7.0]);

    # minimizer 'state'
    my $state = Games::Go::AGA::BayRate::GSL::Multimin::my_fdfminimizer_set (
        $gsl_multimin_fdfminimizer_conjugate_fr,    #minimizer type
        # gsl_multimin_function_fdf structure members:
            \&my_f,     # f       function
            \&my_df,    # df      derivative of f
            \&my_fdf,   # fdf     f and df
            2,          # n       number of free variables
            \@params,   # params  function params passed to f, df, and fdf
        # end of gsl_multimin_function_fdf structure members:
        $x->raw,    # starting point vector
        0.01,       # step size
        1e-4,       # accuracy
     );

    my $iter = 0;
    my $status = $GSL_CONTINUE;

    while(     $status == $GSL_CONTINUE
           and $iter <= 100) {
        $iter++;
        $status = gsl_multimin_fdfminimizer_iterate($state);

        last if ($status);

        my $gradient = gsl_multimin_fdfminimizer_gradient($state);
        $status = gsl_multimin_test_gradient($gradient, 1e-3);

        if ($status == $GSL_SUCCESS) {
            print "\nConverged to minimum at\n";
        }
        my $raw_x = gsl_multimin_fdfminimizer_x($state);
        printf("%5d %.5f %.5f %10.5f\n",
            $iter,
            gsl_vector_get($raw_x, 0),
            gsl_vector_get($raw_x, 1),
          # gsl_multimin_fdfminimizer_f($state),    # hmm, no such function?
                                                    # ok, do it this way instead:
            Games::Go::AGA::BayRate::GSL::Multimin::my_fdfminimizer_f($state),
        );
    }

    if ($status != $GSL_SUCCESS) {
        print "Error: ", gsl_strerror($status), "\n";
    }

    return $status;
}

minimize();


extra/after_build.pl
#!/usr/bin/perl
#===============================================================================
#
#  DESCRIPTION:  Modify Makefile.PL after Dist::Zilla is finished
#
#       AUTHOR:  Reid Augustin
#        EMAIL:  reid@LucidPort.com
#      CREATED:  07/28/2011 03:40:36 PM
#===============================================================================

use strict;
use warnings;
use Carp;
use File::Spec;
use File::Slurp;

my $dir = $ARGV[0] or die "Need build directory";

fix_makefile($dir);
exit 0;

sub fix_makefile {
    my ($dir) = @_;

    my $filename = File::Spec->catfile($dir, 'Makefile.PL');

    my $content = read_file($filename);

    my $add = q{

print "Checking for GNU Scientific Library (GSL)...";

my $libpath;

eval {
    $libpath = `pkg-config --libs gsl 2>/dev/null`;
};
if (not $libpath) {
    eval {
        $libpath = `gsl-config --libs 2>/dev/null`;
    }
};

if (not $libpath) {
    die("I can't seem to find the GSL library (via pkg-config or gsl-config).\n"
      . "Please install the GSL library (using your package manager,\n"
      . "'cpanm Alien::GSL', or compile from source), and try again\n"
    );
}

    };

    $content =~ s/^(use ExtUtils::MakeMaker[^\n]*)/$1\n$add/sm;

    # $content =~ s/("NAME"\s*=>\s*"Games:[^\n]*)/$1\n$add/sm;

    write_file($filename, $content);
}




lib/Games/Go/AGA/BayRate.pm
#*************************************************************************************
#
#     Copyright 2010 Philip Waldron
#
#     This file is part of BayRate.
#
#     BayRate is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     BayRate is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with BayRate.  If not, see <http://www.gnu.org/licenses/>.
#
#**************************************************************************************
#===============================================================================
#
#     ABSTRACT:  implementation of AGA BayRate (player ratings) as perl object
#
#       AUTHOR:  Reid Augustin (REID), <reid@lucidport.com>
#        EMAIL:  reid@LucidPort.com
#      CREATED:  12/02/2010 08:51:22 AM PST
#===============================================================================

use 5.008;
use strict;
use warnings;

package Games::Go::AGA::BayRate;

use base Games::Go::AGA::BayRate::Collection;

our $VERSION = '0.071'; # VERSION

sub new {
    my ($proto, %args) = @_;

    my $self = {};
    bless($self, ref($proto) || $proto);

    my $collection
    = $self->{collection}
    = Games::Go::AGA::BayRate::Collection->new(
            iter_hook             => \&iter_hook, # called once per f or fdf iteration
        #   fdf_iterations        =>  # number fdf_iterations to perform
        #   fdf_gradient_spec     =>  # fdf gradient to test against
        #   f_iterations          =>  # number f_iterations to perform
        #   f_size                =>  # f size to test against
        #   calc_ratings_failover =>  # force failover to calc_ratings_f
        #   calc_sigma_failover   =>  # force failover to calc_sigma2
        #   strict_compliance     =>  # adnere exactly to original bayrate C++ code
    );

    # enter all the players who were in a game
    my $players = $args{players} || [];
    foreach my $player ( @{$players} ) {
        $self->add_player($player);
    }

    # enter all the games
    my $games = $args{games} || [];
    foreach my $game (@{$games}) {
        $self->add_game($game);
    }
    return $self;
}

# hook called for each F(DF)Minimzer iteration
sub iter_hook {
    my ($collection, $state, $iter, $status) = @_;

    if (ref $state eq 'Math::GSL::Multimin::gsl_multimin_fminimizer') {
        my $f = # gsl_multimin_fminimizer_fval($state),    # hmm, struct member, not a function
                # ok, do it this way instead:
            Games::Go::AGA::BayRate::GSL::Multimin::my_fminimizer_fval($state),
        my $size = gsl_multimin_fminimizer_size($state);
        printf("F Iteration %d\tf() = %g\tsimplex size = %g\n", $iter, $f, $size);
        if ($status == $GSL_SUCCESS) {
            printf "\nConverged to minimum. f() = %g\n", $f;
        }
    }
    elsif (ref $state eq 'Math::GSL::Multimin::gsl_multimin_fdfminimizer') {
        my $gradient = Games::Go::AGA::BayRate::GSL::Multimin::my_fdfminimizer_gradient($state);
        my $minimum = gsl_multimin_fdfminimizer_minimum($state);
        printf("FDF Iteration %d\tf() = %g\tnorm = %g\tStatus = %d\n",
            $iter,
            $minimum,
            gsl_blas_dnrm2($gradient),
            $status);
        if ($status == $GSL_SUCCESS) {
            printf "\nConverged to minimum. Norm(gradient) = %g\n",
                gsl_blas_dnrm2($gradient),
        }

    }
    else {
        die(sprintf("Unknown minimizer state type: %s", ref $state));
    }
}

sub add_player {
    my ($self, $player) = @_;

    $self->{collection}->add_player(
        id     => $player->id,
        seed   => $player->rating,
        #  sigma  => 6.0,  # will get changed later
    );
}

sub add_game {
    my ($self, $game) = @_;

    return if (not $game->winner);  # skip games without results
    $self->{collection}->add_game(
        white     => $game->white.
        black     => $game->black.
        whiteWins => $game->winner->id eq $game->white->id,
        handicap  => $game->handicap,
        komi      => $game->komi,
    );
}

1;

__END__

=head1 SYNOPSIS

    use Games::Go::AGA::BayRate;

    my $bayrate = Games::Go::AGA::BayRate->new(
        players => $tournament->players,    # ref to array of players
        games   => $tournament->games,      # ref to array of games
    };

    # assuming we don't have carryover sigma info:
    $bayrate->initSeeding($tdList);

    # caclulate new ratings
    $bayrate->calc_ratings;

    for (my $bay_player (@${bayrate->players}) {
        # transfer adjusted ratings back to tournament players
        my $tourn_player = $tournament->find_player($bay_player->id);
        $tourn_player->set_adj_rating($bay_player->rating);
    }

=head1 DESCRIPTION

The American Go Association (AGA) provides a rating system for the go
players of the nation.  The algorithm is described in detail in a paper
on their web-site (http://usgo.org):  AGARatings-Math.pdf.  They also
provide a C++ implementation example: C<bayrate.zip>.

This package implements a perl version of C<bayrate>, both the
executable and the support objects (Game, Player, etc).  Note: only
C<bin/bayrate> is included here, C<singlerate> and C<check> are left as
an excersise for the student.

C<bayrate>, and this module, require a fairly recent version of the GNU
Scientific Library (GSL), including the devel portion (containing header
files, etc).  If you are using Fedora (linux), "yum install gsl
gsl-devel" should be sufficient.  Version 1.14 and later should work,
earlier versions may not.  To find your current version, run:

    pkg-config --modversion gsl

Work is being done to provide a full perl interface to GSL, but it is
not complete as of this writing.  I have used Inline::C to hook to the
specific GSL functions used by bayrate.pl.  The missing parts needed by
bayrate.pl are the Multimin functions (f_minimizer and fdf_minimizer).
Tests for the Inline::C interface to these functions (as well as the C
versions as called out in the Gnu GSL documentation) are included in the
'extra' subdirectory of this package.

This module interfaces between a C<Games::Go::AGA::DataObjects> objects
and a C<Games::Go::AGA::BayRate::Collection> object.  A new BayRate
object subclasses the Collection object and copies information from a
list of C<Games::Go::AGA::DataObjects::Players> and a list of
C<Games::Go::AGA::DataObjects::Games> (such as the lists provided by the
C<players> and C<games> methods of a
C<Games::Go::AGA::DataObjects::Tournament> object) to the Collection.
After calling the C<$bayrate-E<gt>calc_ratings> method to run
the C<Collection> rating algorithm,  The adjusted ratings are available
to copy back into the C<adj_rating>s of the Tournament players.

=head1 IMPLEMENTATION

Given a collection of players, each with a rating and a sigma, and given
a collection of games the players may or may not have participated in
(including the outcomes of those games), this module uses Bayesian
statistical mathods to determine an adjusted rating for each player so
that the observed outcomes are considered to be 'most likely'.

In practice, what this means is that if player A defeats player B, but
because of the difference in rating, this is the expected outcome, then
the adjustments to the ratings of player A and player B will be rather
small.  On the other hand, if player B is expected to defeat player A,
then the rating of player A should be raised rather significantly, and
player B should be correspondingly lowered.

To use this in a tournament, enter the players and all the games after
each round (NOTE: use the players' B<initial> ratings, not their
adjusted ratings, and add B<all> the games from B<all> rounds).  For
example, after round 3 is complete, enter all the games from rounds 1,
2, and 3.  Then use the adjusted ratings to help find pairings for the
next round (See Games::Go::W3gtd using the Best pairing method).

The effects of wins and losses propagate so that the adjustment to
player A (who played against player B in the first round) can be
affected by the outcome of player B's game against player C (in a
subsequent round).

=head1 SEE ALSO

=over

=item Games::Go::AGA::BayRate::Collection

=item Games::Go::AGA::BayRate::Game

=item Games::Go::AGA::BayRate::Player

=item Games::Go::W3gtd

=back



lib/Games/Go/AGA/BayRate/Game.pm
#*************************************************************************************
#
#     Copyright 2010 Philip Waldron
#
#     This file is part of BayRate.
#
#     BayRate is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     BayRate is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with BayRate.  If not, see <http://www.gnu.org/licenses/>.
#
#***************************************************************************************
#===============================================================================
#
#     ABSTRACT:  re-factored game.cpp portion of AGA BayRate in perl
#
#       AUTHOR:  Reid Augustin
#        EMAIL:  reid@LucidPort.com
#      CREATED:  05/24/2011 11:05:53 AM
#===============================================================================

use 5.008;
use strict;
use warnings;

package Games::Go::AGA::BayRate::Game;

use Readonly;
use Carp;
use Math::GSL::SF ( ':all' );   # Special Functions (for sf_erfc)

our $VERSION = '0.071'; # VERSION

Readonly our @attr => (
    'komi',        # Komi
    'handicap',    # Actual handicap
    'whiteWins',   # True if White wins
    'white',       # white player
    'black',       # black player
);

sub new {
    my ($proto, %args) = @_;

    my $self = {};
    bless($self, ref($proto) || $proto);
    foreach my $name (@attr) {
        # transfer user args
        if (not exists($args{$name})) {
            croak("$name not defined");
        }
        $self->{$name} = delete $args{$name};
    }
    if (keys %args) {
        croak sprintf "Unknown argument: %s", join(', ', keys %args);
    }
    my $handicap = $self->get_handicap;
    my $komi = $self->get_komi;
    if ($handicap > 9 or
        $handicap < 0) {
        croak "Handicap $handicap out of range (0-9)";
    }
    if ($handicap >=2 and
        ($komi >= 10 or
         $komi <= -10)) {
        croak "Komi($komi) for handicap $handicap must be in range (-10, 10)";
    }
    if ($handicap <= 1 and
        ($komi <= -20 or
         $komi >= 20)) {
        croak "Komi($komi) for handicap $handicap must be in range (-20, 20)";
    }
    return($self);
}

# make read accessor methods
foreach my $name (@attr) {
    no strict 'refs'; ## no critic
    *{"get_$name"} = sub {
        return shift->{$name};
    }
}

# these read accessor methods need a little extra
foreach my $name ('sigma_px',
                  'handicapeqv', # Equivalent rating adjustment based on handicap and komi
                ) {
    no strict 'refs'; ## no critic
    *{"get_$name"} = sub {
        my ($self) = @_;

        $self->calc_handicapeqv if (not exists($self->{$name}));
        return $self->{$name};
    }
}

# make write accessor methods
foreach my $name (qw(
            whiteWins
            komi
            handicap
        )) {
    no strict 'refs'; ## no critic
    *{"set_$name"} = sub {
        my ($self, $val) = @_;
        $self->{$name} = $val;
        delete $self->{sigma_px};
        delete $self->{handicapeqv};
        return shift->{$name};
    }
}

#****************************************************************
#
# calc_handicapeqv ()
#
# Calculates the equivalent rating difference for the handicap/komi
# conditions of this game.
#
#****************************************************************
Readonly our @sigma_add_from_handicap => (
    undef,
    undef,
    1.13672,    # case 2
    1.18795,    # case 3
    1.22841,    # case 4
    1.27457,    # case 5
    1.31978,    # case 6
    1.35881,    # case 7
    1.39782,    # case 8
    1.43614,    # case 9
);

sub calc_handicapeqv {
    my ($self) = @_;

    if ($self->{handicap} == 0 or
        $self->{handicap} == 1) {
        $self->{handicapeqv} = 0.580 - 0.0757 * $self->{komi};
        $self->{sigma_px} = 1.0649 - 0.0021976 * $self->{komi} + 0.00014984 * $self->{komi} * $self->{komi};
    }
    else {
        $self->{handicapeqv} = $self->{handicap} - 0.0757 * $self->{komi};
        if (not defined $sigma_add_from_handicap[$self->{handicap}]) {
            croak "handicap $self->{handicap} out of range"
        }
        $self->{sigma_px} = -0.0035169 * $self->{komi};
        $self->{sigma_px} += $sigma_add_from_handicap[$self->{handicap}]
    }
    croak "sigma_px is $self->{sigma_px}" if ($self->{sigma_px} <= 0);
}

Readonly our $SQRT_OF_TWO => sqrt(2.0);
# return probability of this game result based on players' seeds
sub seed_probability {
    my ($self) = @_;

    my $white = $self->get_white;
    my $black = $self->get_black;

    my $rd = $white->get_cseed - $black->get_cseed - $self->get_handicapeqv;

    if ($self->get_whiteWins) {
        $rd = -$rd;
    }
    return gsl_sf_erfc($rd / $self->get_sigma_px / $SQRT_OF_TWO) / 2.0;
}

1;

=head1 SYNOPSIS

  use Games::Go::AGA::BayRate::Game;

=head1 DESCRIPTION

Games::Go::AGA::BayRate::Game is a perl implementation of game.cpp
found in the C<bayrate.zip> package from the American Go Association
(http://usgo.org).

Most of the following documentation is shamelessly stolen directly from
C<game.cpp>.

=head1 METHODS

=over

=item get_komi ()

=item get_handicap ()

=item get_whiteWins ()

=item set_komi ( new )

=item set_handicap ( new )

=item set_whiteWins ( new )

=item get_white ()

=item get_black ()

=item get_sigma_px ()

=item get_handicapeqv ()

Accessors for getting and setting game attributes.  Attributes without a
'C<set_>' method must be created at C<->new(...)> time.

=item calc_handicapeqv ()

Calculates the equivalent rating difference for the handicap/komi
conditions of this game.  C<get_sigma_px()> and C<get_handicapeqv()>
call this if necessary.

=item seed_probability

Returns the probability of this game result based on players' seeds

=back


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;


print "Checking for GNU Scientific Library (GSL)...";

my $libpath;

eval {
    $libpath = `pkg-config --libs gsl 2>/dev/null`;
};
if (not $libpath) {
    eval {
        $libpath = `gsl-config --libs 2>/dev/null`;
    }
};

if (not $libpath) {
    die("I can't seem to find the GSL library (via pkg-config or gsl-config).\n"
      . "Please install the GSL library (using your package manager,\n"
      . "'cpanm Alien::GSL', or compile from source), and try again\n"
    );
}

    



my %WriteMakefileArgs = (
  "ABSTRACT" => "implementation of AGA BayRate (player ratings) as perl object",
  "AUTHOR" => "Reid Augustin <reid\@hellosix.com>",
  "BUILD_REQUIRES" => {
    "Devel::CheckLib" => 0,
    "English" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Games-Go-AGA-BayRate",
  "EXE_FILES" => [
    "bin/bayrate.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Games::Go::AGA::BayRate",
  "PREREQ_PM" => {
    "Carp" => 0,
    "DBI" => 0,
    "Date::Parse" => 0,
    "DateTime" => 0,
    "Getopt::Long" => 0,
    "IO::File" => 0,
    "Inline" => 0,
    "Math::GSL::BLAS" => 0,
    "Math::GSL::Const" => 0,
    "Math::GSL::Errno" => 0,
    "Math::GSL::Interp" => 0,
    "Math::GSL::Linalg" => 0,
    "Math::GSL::Matrix" => 0,
    "Math::GSL::Multimin" => 0,
    "Math::GSL::Permutation" => 0,
    "Math::GSL::SF" => 0,
    "Math::GSL::Spline" => 0,
    "Math::GSL::Vector" => 0,
    "Readonly" => 0,
    "Scalar::Util" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.071",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



